### DISCUSSION TITLE
Middleware example for 4.0.0

### DISCUSSION THREAD

**Discussion by GreenTeaSeb** [actix/actix-web#2597] [Q&A] [answered]
I've been trying to add some middleware to my app in the beta version, but with the changes to Transform and Service the example in the docs is no longer helping me.

I've looked at the new docs for the traits, I've managed to fix it but get stuck at the call function.
I need to redirect the user to a page but I keep getting the error

```rust
    the trait `std::convert::From<&BoxBody>` is not implemented for `actix_web::dev::Response<_>
```
```rust
fn call(&self, req: ServiceRequest) -> Self::Future {
        let is_logged_in = false;

        if is_logged_in {
            Either::Left(self.service.call(req))
        }else {
	    Either::Right(ok(req.into_response(
		HttpResponse::Ok().finish().body()
	    )
	    ))
     }
}
```

I also noticed that actix-web now has Either, not sure if that replaces the futures::Either as well.
Could anyone provide an example?

**Comment by aliemjay** [comment, marked as answer]
You'll need `EitherBody` and `ServiceResponse::{map_into_right_body, map_into_left_body}`

```rust
type Response = ServiceResponse<EitherBody<B, BoxBody>>;

fn call() {
  let service = Rc::clone(&self.service);
  Box::pin(async move {
    if true {
      Ok(service.call(req)?.await.map_into_left_body())
    } else {
      let resp = HttpResponse::Ok().finish();
      Ok(req.into_response(resp).map_into_right_body())
    }
  })
}
```

alternatively you can box the response body: 

```rust
type Response = ServiceResponse<BoxBody>; // or `ServiceResponse`

fn call() {
  let service = Rc::clone(&self.service);
  Box::pin(async move {
    if true {
      Ok(service.call(req)?.await.map_into_boxed_body())
    } else {
      let resp = HttpResponse::Ok().finish();
      Ok(req.into_response(resp).map_into_boxed_body())
    }
  })
}
```

**Reply by GreenTeaSeb** [reply]
Figured out the 2nd method on my own. But this is better than the way I did it, so I'll go with those.
Thank you so much!

**Reply by lizhongit** [reply]
I'm migration that from version 3 to verion 4, and change 

I got an errors

```shell
24  |       .wrap(middleware::SayHi)
    |        ---- ^^^^^^^^^^^^^^^^^ expected enum `EitherBody`, found struct `BoxBody`
```

Could you show me the full code?
Thank you so mush!

**Comment by lermana** [comment]
In case this helps anyone else who stumbles here, `actix_web_lab` (which seems to serve as a proving ground for potential `actix_web` features) provides `from_fn`, which allows for writing standalone, `async` functions for middleware ([docs](https://docs.rs/actix-web-lab/latest/actix_web_lab/middleware/fn.from_fn.html)). This lets you do things like:
```
pub async fn authz_mw(
    req: ServiceRequest,
    next: Next<impl MessageBody>,
) -> Result<ServiceResponse<impl MessageBody>, Error> {
    let is_authz = authz(&req).await;

    match is_authz {
        true => Ok(next.call(req).await?),
        false => Err(actix_web::error::ErrorForbidden("Unauthorized to perform requested action"))
    }
}

...
App::new()
    ...
    // middleware
    .wrap(from_fn(authz_mw)) 
```

Really helpful capability!

**Comment by jsoneaday** [comment]
This worked for me but I made my B bound by MessageBody. This is causing to infinite loop on poll_next. Anyone seeing this?

**Comment by ejaznoob** [comment]
None of the comments here works for me to implement a simple middleware for actix_web 4. Not even a single working example exists on whole internet.

**Reply by robjtede** [reply]
Please see the official middleware examples here: https://github.com/actix/examples/tree/master/middleware
